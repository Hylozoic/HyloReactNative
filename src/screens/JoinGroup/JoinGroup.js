import React, { useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useFocusEffect, useNavigation, useRoute } from '@react-navigation/native'
import { every, isEmpty } from 'lodash/fp'
import setReturnToOnAuthPath from 'store/actions/setReturnToOnAuthPath'
import getRouteParam from 'store/selectors/getRouteParam'
import { getAuthorized } from 'store/selectors/getAuthState'
import { openURL } from 'hooks/useOpenURL'
import useInvitation from 'store/actions/useInvitation'
import checkInvitation from 'store/actions/checkInvitation'
import LoadingScreen from 'screens/LoadingScreen'

export const SIGNUP_PATH = '/signup'
export const EXPIRED_INVITE_PATH = '/invite-expired'

export default function JoinGroup (props) {
  const navigation = useNavigation()
  const route = useRoute()
  const dispatch = useDispatch()
  const isAuthorized = useSelector(getAuthorized)

  // Might be more clear to simply use `useEffect`
  useFocusEffect(
    useCallback(() => {
      (async function () {
        try {
          const invitationTokenAndCode = {
            invitationToken: getRouteParam('token', route),
            accessCode: getRouteParam('accessCode', route)
          }

          if (every(isEmpty, invitationTokenAndCode)) {
            throw new Error('Please provide either a `token` query string parameter or `accessCode` route param')
          }

          if (isAuthorized) {
            // eslint-disable-next-line react-hooks/rules-of-hooks
            const result = await dispatch(useInvitation(invitationTokenAndCode))
            const newMembership = result?.payload?.getData()?.membership
            const groupSlug = newMembership?.group?.slug

            if (groupSlug) {
              openURL(`/groups/${groupSlug}`, true)
            } else {
              throw new Error('Join group was unsuccessful')
            }
          } else {
            const result = await dispatch(checkInvitation(invitationTokenAndCode))
            const isValidInvite = result?.payload?.getData()?.valid

            if (isValidInvite) {
              dispatch(setReturnToOnAuthPath(route.params?.originalLinkingPath))
              openURL('/signup?message=Signup or login to join this group.', true)
            } else {
              openURL('/signup?error=invite-expired', true)
            }
          }
        } catch (error) {
          console.log('!!! error', error)
          navigation.canGoBack() ? navigation.goBack() : openURL('/')
        }
      })()
    }, [])
  )

  return <LoadingScreen />
}
