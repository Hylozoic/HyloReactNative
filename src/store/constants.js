/**
 * @providesModule store/constants
 */

export const CHANGE_COMMUNITY = 'CHANGE_COMMUNITY'
export const EXTRACT_MODEL = 'EXTRACT_MODEL'
export const SET_STATE = 'SET_STATE'
export const LOG_ERROR = 'LOG_ERROR'
