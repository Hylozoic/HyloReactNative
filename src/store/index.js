import { applyMiddleware, createStore } from 'redux'
import { compact } from 'lodash'
import rootReducer from '../reducer'
import promiseMiddleware from 'redux-promise'
import apiMiddleware from './middleware/api'
import graphQLMiddleware from './middleware/graphQL'
import { createLogger } from 'redux-logger'
import { AsyncStorage } from 'react-native'
import { PERSISTED_STATE_KEY } from '../reducer/persistence'

const middleware = compact([
  graphQLMiddleware,
  apiMiddleware,
  promiseMiddleware,
  __DEV__ && createLogger({collapsed: true})
])

function getInitialState () {
  return AsyncStorage.getItem(PERSISTED_STATE_KEY)
  .then(state => state ? JSON.parse(state) : {})
}

export default function getStore () {
  return getInitialState().then(initialState => {
    const store = createStore(rootReducer, initialState,
      applyMiddleware(...middleware))

    // Enable Webpack hot module replacement for reducers
    if (module.hot) {
      module.hot.accept(() => {
        const nextRootReducer = require('../reducer')
        store.replaceReducer(nextRootReducer)
      })
    }

    return store
  })
}
