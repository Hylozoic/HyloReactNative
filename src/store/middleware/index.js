import { createLogger } from 'redux-logger'
import { compact } from 'lodash'
import afterInteractionsMiddleware from './afterInteractions'
import apiMiddleware from './api'
import graphQLMiddleware from './graphQL'
import optimisticMiddleware from './optimistic'
import pendingMiddleware from './pending'
import promiseMiddleware from 'redux-promise'

const middleware = compact([
  graphQLMiddleware,
  apiMiddleware,
  afterInteractionsMiddleware,
  optimisticMiddleware,
  pendingMiddleware,
  promiseMiddleware,
  __DEV__ && createLogger({
    collapsed: (getState, action, logEntry) => !logEntry.error
  })
])

export default middleware
