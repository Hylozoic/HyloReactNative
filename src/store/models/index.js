import { ORM } from 'redux-orm'
import './Model.extension'
import Activity from './Activity'
import Attachment from './Attachment'
import Comment from './Comment'
import EventInvitation from './EventInvitation'
import Group, {
  GroupRelationship, GroupModerator, GroupJoinQuestion,
  GroupPrerequisite, GroupToGroupJoinQuestion
} from './Group'
import GroupRelationshipInvite, { GroupToGroupJoinRequestQuestionAnswer } from './GroupRelationshipInvite'
import GroupTopic from './GroupTopic'
import Invitation from './Invitation'
import Location from './Location'
import Me, { MySkillsToLearn } from './Me'
import Membership from './Membership'
import Person, { PersonSkillsToLearn } from './Person'
import PersonConnection from './PersonConnection'
import Message from './Message'
import MessageThread from './MessageThread'
import Notification from './Notification'
import Post, { PostFollower, PostCommenter, ProjectMember } from './Post'
import PostMembership from './PostMembership'
import SearchResult from './SearchResult'
import Skill from './Skill'
import Topic from './Topic'
import Vote from './Vote'
import JoinRequest, { JoinRequestQuestionAnswer, Question } from './JoinRequest'

export const orm = new ORM({ stateSelector: state => state.orm })

orm.register(
  Activity,
  Attachment,
  Comment,
  EventInvitation,
  Group,
  GroupJoinQuestion,
  GroupModerator,
  GroupPrerequisite,
  GroupRelationship,
  GroupRelationshipInvite,
  GroupToGroupJoinQuestion,
  GroupToGroupJoinRequestQuestionAnswer,
  GroupTopic,
  Invitation,
  JoinRequest,
  JoinRequestQuestionAnswer,
  Location,
  Me,
  Membership,
  Message,
  MessageThread,
  MySkillsToLearn,
  Notification,
  Person,
  PersonConnection,
  PersonSkillsToLearn,
  Post,
  PostCommenter,
  PostFollower,
  PostMembership,
  ProjectMember,
  Question,
  SearchResult,
  Skill,
  Topic,
  Vote
)

export default orm
