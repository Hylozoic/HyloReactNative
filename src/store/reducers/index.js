import { combineReducers } from 'redux'
import resetStore from './resetStore'
import ormReducer from './ormReducer'
import pending from './pending'
import queryResults from './queryResults'
import currentNetworkAndCommunity from './currentNetworkAndCommunity'
import sessionReducer from './sessionReducer'
import mixpanelReducer from './mixpanel'
import CommentEditor from '../../components/PostDetails/CommentEditor/CommentEditor.store'
import DeepLinkHandler from '../../components/DeepLinkHandler/DeepLinkHandler.store'
import FeedList from '../../components/FeedList/FeedList.store'
import ItemChooser from '../../components/ItemChooser/ItemChooser.store'
import LoadingModal from '../../components/LoadingModal/LoadingModal.store'
import MemberFeed from '../../components/MemberProfile/MemberFeed/MemberFeed.store'
import Members from '../../components/Tabs/Members/Members.store'
import ModeratorSettings from '../../components/ModeratorSettings/ModeratorSettings.store'
import NewMessage from '../../components/NewMessage/NewMessage.store'
import PeopleTyping from '../../components/PeopleTyping/PeopleTyping.store'
import PostEditor from '../../components/PostEditor/PostEditor.store'
import SkillEditor from '../../components/SkillEditor/SkillEditor.store'
import SignupFlow from '../../components/SignupFlow/SignupFlow.store'
import SocketListener from '../../components/SocketListener/SocketListener.store'
import TabBar from '../../components/RootNavigator/TabBar/TabBar.store'
import Topics from '../../components/Tabs/Topics/Topics.store'
import CreateCommunityFlow from '../../components/CreateCommunityFlow/CreateCommunityFlow.store'
import SearchPage from '../../components/SearchPage/SearchPage.store'

export const combinedReducers = combineReducers({
  orm: ormReducer,
  pending,
  queryResults,
  currentNetworkAndCommunity,
  session: sessionReducer,
  mixpanel: mixpanelReducer,
  CommentEditor,
  DeepLinkHandler,
  FeedList,
  ItemChooser,
  LoadingModal,
  MemberFeed,
  Members,
  ModeratorSettings,
  NewMessage,
  PeopleTyping,
  PostEditor,
  SearchPage,
  SignupFlow,
  SkillEditor,
  SocketListener,
  TabBar,
  Topics,
  CreateCommunityFlow
})

const composeReducers = (...reducers) => (state, action) =>
  reducers.reduce((newState, reducer) => reducer(newState, action), state)

export default composeReducers(
  combinedReducers,
  resetStore
)
