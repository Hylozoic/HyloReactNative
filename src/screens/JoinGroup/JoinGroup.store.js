import { get } from 'lodash/fp'
import { AnalyticsEvents } from 'hylo-utils/constants'

export const MODULE_NAME = 'JoinGroup'
export const USE_INVITATION = `${MODULE_NAME}/USE_INVITATION`
export const CHECK_INVITATION = `${MODULE_NAME}/CHECK_INVITATION`

export function useInvitation (userId, invitationCodes = {}) {
  const { invitationToken, accessCode } = invitationCodes
  return {
    type: USE_INVITATION,
    graphql: {
      query: `mutation ($userId: ID, $invitationToken: String, $accessCode: String) {
        useInvitation (userId: $userId, invitationToken: $invitationToken, accessCode: $accessCode) {
          membership {
            id
            role
            group {
              id
              name
              slug
            }
          }
          error
        }
      }`,
      variables: {
        userId,
        invitationToken,
        accessCode
      }
    },
    meta: {
      extractModel: [
        {
          modelName: 'Membership',
          getRoot: get('useInvitation.membership')
        }
      ],
      analytics: AnalyticsEvents.GROUP_INVITATION_ACCEPTED
    }
  }
}

export function checkInvitation (invitationCodes) {
  const { invitationToken, accessCode } = invitationCodes
  return {
    type: CHECK_INVITATION,
    graphql: {
      query: `query ($invitationToken: String, $accessCode: String) {
        checkInvitation (invitationToken: $invitationToken, accessCode: $accessCode) {
          valid
        }
      }`,
      variables: {
        invitationToken,
        accessCode
      }
    }
  }
}
