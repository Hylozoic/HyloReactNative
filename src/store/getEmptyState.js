import orm from './models'
import { combinedReducers } from './reducers'

export default function getEmptyState () {
  return combinedReducers({ orm: orm.getEmptyState() }, { type: '' })
}
