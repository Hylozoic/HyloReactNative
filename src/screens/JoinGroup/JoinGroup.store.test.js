import { useInvitation } from './JoinGroup.store'

test('useInvitation', () => {
  const invitationCodes = {
    invitationToken: 'invitationtoken',
    accessCode: 'accesstoken'
  }
  expect(useInvitation(invitationCodes)).toMatchSnapshot()
})
