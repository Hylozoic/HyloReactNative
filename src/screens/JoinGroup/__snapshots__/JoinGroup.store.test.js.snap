// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useInvitation 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($invitationToken: String, $accessCode: String) {
        useInvitation (invitationToken: $invitationToken, accessCode: $accessCode) {
          membership {
            id
            role
            group {
              id
              accessibility
              name
              settings {
                allowGroupInvites
                askJoinQuestions
                askGroupToGroupJoinQuestions
                publicMemberDirectory
                showSuggestedSkills
              }
              slug
              visibility
            }
            person {
              id
            }
            settings {
              showJoinForm
            }
          }
          error
        }
      }",
    "variables": Object {
      "accessCode": "accesstoken",
      "invitationToken": "invitationtoken",
    },
  },
  "meta": Object {
    "analytics": "Group Invitation Accepted",
    "extractModel": Object {
      "getRoot": [Function],
      "modelName": "Membership",
    },
  },
  "type": "JoinGroup/USE_INVITATION",
}
`;
